src/library/basesqltablemodel.cpp

```diff
@@ -419,19 +375,102 @@ void BaseSqlTableModel::setSort(int column, Qt::SortOrder order) {
     }
 
-    bool sortColumnChanged = m_iSortColumn != column;
-    bool sortOrderChanged = m_eSortOrder != order;
+    int trackSourceColumnCount = m_trackSource ? m_trackSource->columnCount() : 0;
 
-    if (!sortColumnChanged && !sortOrderChanged) {
-        // Do nothing if the sort is not different.
+    if (column < 0 ||
+            column >= trackSourceColumnCount + m_sortColumns.size() - 1) {
+        // -1 because id column is in both tables
+        qWarning() << "BaseSqlTableModel::setSort invalid column:" << column;
         return;
     }
 
-    // TODO(rryan) optimization: if the sort column has not changed but the
-    // order has, just reverse our ordering of the rows.
+    if (m_sortColumns.size() > 0 &&
+            m_sortColumns.at(0).column == column) {
+        // Only the order has changed
+        SortColumn sc;
+        sc.column = column;
+        sc.order = order;
+        m_sortColumns.replace(0, sc);
+    } else {
+        // remove column if already in history
+        for (int i = 0; i < m_sortColumns.size(); ++i) {
+            if (m_sortColumns.at(i).column == column) {
+                m_sortColumns.removeAt(i);
+                break;
+            }
+        }
+
+        // set new sort as head and shift out old sort
+        SortColumn sc;
+        sc.column = column;
+        sc.order = order;
+        m_sortColumns.prepend(sc);
+
+        if (m_sortColumns.size() > 3) {
+            m_sortColumns.removeLast();
+        }
+    }
 
-    m_iSortColumn = column;
-    m_eSortOrder = order;
+    // we have two selects for sorting, since keeping the select history
+    // across the two selects is hard, we do this only for the trackSource
+    // this is OK, because the coloums of the table are virtual in cas of
+    // preview column or individual like playlist track number so that we
+    // do not need the history anyway.
+
+    // reset the old order by clauses
+    m_trackSourceOrderBy.clear();
+    m_tableOrderBy.clear();
+    m_trackSourceSortColumn = 0;
+    m_trackSourceSortOrder = Qt::AscendingOrder;
+
+    if (column > 0 && column < m_tableColumns.size()) {
+        // Table sorting, no history
+        m_tableOrderBy.append(QLatin1String("ORDER BY "));
+        QString field = m_tableColumns[column];
+        QString sort_field = QString("%1.%2").arg(m_tableName, field);
+        m_tableOrderBy.append(sort_field);
+    #ifdef __SQLITE3__
+        m_tableOrderBy.append(QLatin1String(" COLLATE localeAwareCompare"));
+    #endif
+        m_tableOrderBy.append((order == Qt::AscendingOrder) ?
+                QLatin1String(" ASC") :
+                QLatin1String(" DESC"));
+        m_sortColumns.clear();
+    } else if (m_trackSource) {
+        foreach (SortColumn sc, m_sortColumns) {
+            // TrackSource Sorting, current sort + two from history
+            if (!m_trackSourceOrderBy.isEmpty()) {
+                // second cycle
+                m_trackSourceOrderBy.append(QLatin1String(", "));
```

Comment:
is QLatin1String really necessary for a comma?  literal seems fine since this is a database clause, not a translated numeral separator or something.